kasa/tapo/tapodevice.py

```diff
@@ -28,38 +30,67 @@ def __init__(
             transport=AesTransport(config=config or DeviceConfig(host=host)),
         )
         super().__init__(host=host, config=config, protocol=_protocol)
-        self._components: Optional[Dict[str, Any]] = None
+        self._components_raw: Optional[Dict[str, Any]] = None
+        self._components: Dict[str, int]
         self._state_information: Dict[str, Any] = {}
         self._discovery_info: Optional[Dict[str, Any]] = None
+        self.modules: Dict[str, Any] = {}
 
     async def update(self, update_children: bool = True):
         """Update the device."""
         if self.credentials is None or self.credentials.username is None:
             raise AuthenticationException("Tapo plug requires authentication.")
 
-        if self._components is None:
+        if self._components_raw is None:
             resp = await self.protocol.query("component_nego")
-            self._components = resp["component_nego"]
+            self._components_raw = resp["component_nego"]
+            self._components = {
+                comp["id"]: comp["ver_code"]
+                for comp in self._components_raw["component_list"]
+            }
+            await self._initialize_modules()
+
+        extra_reqs: Dict[str, Any] = {}
+        if "energy_monitoring" in self._components:
+            extra_reqs = {
+                **extra_reqs,
+                "get_energy_usage": None,
+                "get_current_power": None,
+            }
```

Comment:
Maybe initialize_modules should build the final query to be used for updates?

Reply:
Agreed. See my comment above wrt. adding module support. We can re-use the already existing interface with a tapodevice specific implementation.